File: Atcoder/abc180/E.cpp
/*	2021-06-19	 */
#include <bits/stdc++.h>
#define all(s) (s).begin(),(s).end()
#define rall(s) (s).rbegin(),(s).rend()
#define uni(s) (s).erase(unique(all(s)),(s).end())
using namespace std;
typedef long long ll;
typedef long double ld;
typedef pair<int,int> ii;
typedef tuple<ll,ll,ll> tup;
const ll INF = (ll)(1e15) + 10;
vector<tup> P;
int N;


ll cost(int i,int j){
	return abs(get<0>(P[i]) - get<0>(P[j])) + abs(get<1>(P[i]) - get<1>(P[j])) + max(0LL,get<2>(P[j]) - get<2>(P[i])); 
}

int main(){
	ios::sync_with_stdio(0); cin.tie(nullptr); cout.tie(nullptr);
	cin >> N;
	
	P.resize(N);
	for (int i = 0; i < N; i++){
		int x, y, z;
		cin >> x >> y >> z;
		P[i] = make_tuple(x,y,z);
	}
	
	vector<vector<ll>> dp(N,vector<ll> (1 << N,INF));
	
	dp[0][1 << 0] = 0;
	
	int max_mask = (1 << N) - 1;
	
	
	for (int mask = 0; mask <= max_mask; mask++){
		vector<int> visited;
		vector<int> unvisited;
		for (int i = 0; i < N; i++)
			if ((mask >> i) & 1) visited.push_back(i);
			else unvisited.push_back(i);
		
		if (visited.empty() || unvisited.empty()) continue;
		for (int u: visited)
			for (int v: unvisited)
				dp[v][mask | (1 << v)] = min(dp[v][mask | (1 << v)],dp[u][mask] + cost(u,v));
	}
	
	ll ans = INF;
	for (int i = 1; i < N; i++) 
		ans = min(ans,dp[i][max_mask] + cost(i,0));
	
	cout << ans;
	return 0;
}

File: Atcoder/abc180/E.py
# dp bitmask

N = int(input())
P = []
for _ in range(N):
	xi,yi,zi = map(int,input().split())
	P.append((xi,yi,zi))


def cost(i,j):
	global P
	#print(i,j,P[i],P[j])
	return abs(P[i][0] - P[j][0]) + abs(P[i][1] - P[j][1]) + max(0,P[j][2] - P[i][2])

if __name__ == '__main__':
	
	INF = 10**15
	dp = [[INF for _ in range(1 << N)] for _ in range(N)]
	
	# for i in range(N):
		# dp[i][1 << i] = 0
	dp[0][1 << 0] = 0
	
	for mask in range(1 << N):
		bit1 = []
		bit0 = []
		for i in range(N):
			if (mask >> i) & 1: bit1.append(i)
			else: bit0.append(i)
		
		if len(bit1) == 0 or len(bit0) == 0: continue
		
		for u in bit1:
			for v in bit0:
				dp[v][mask | (1 << v)] = min(dp[u][mask] + cost(u,v),dp[v][mask | (1 << v)])
	
	
	ans = INF
	for i in range(N):
		ans = min(ans,dp[i][(1 << N) - 1] + cost(i,0))
	
	print(ans)
	
	
